Configuration system for CLI (#6708)
qmk/qmk_firmware: docs/cli_development.md

# QMK CLI Development

This document has useful information for developers wishing to write new `qmk` subcommands.

# Overview

> The QMK CLI operates using the subcommand pattern made famous by git. The main `qmk` script is simply there to setup the environment and pick the correct entrypoint to run. Each subcommand is a self-contained module with an entrypoint (decorated by `@cli.subcommand()`) that performs some action and returns a shell returncode, or None.

Reviewer: ```suggestion
The QMK CLI operates using the subcommand pattern made famous by git. The main `qmk` script is simply there to set up the environment and pick the correct entry point to run. Each subcommand is a self-contained module with an entry point (decorated by `@cli.subcommand()`) that performs some action and returns a shell return code, or None.
```

# Subcommands

[MILC](https://github.com/clueboard/milc) is the CLI framework `qmk` uses to handle argument parsing, configuration, logging, and many other features. It lets you focus on writing your tool without wasting your time writing glue code.

Subcommands in the local CLI are always found in `qmk_firmware/lib/python/qmk/cli`.

Let's start by looking at an example subcommand. This is `lib/python/qmk/cli/hello.py`:

```python
"""QMK Python Hello World

This is an example QMK CLI script.
"""
from milc import cli

@cli.argument('-n', '--name', default='World', help='Name to greet.')
@cli.subcommand('QMK Hello World.')
def hello(cli):
    """Log a friendly greeting.
    """
    cli.log.info('Hello, %s!', cli.config.hello.name)
```

> First we import the `cli` object from `milc`. This is how we interact with the user and control the script's behavior. We use `@cli.argument()` to define a command line flag, `--name`. This also creates a configuration variable named `hello.name` (and the corresponding `user.name`) which the user can set so they don't have to specify the argument. The `cli.subcommand()` decorator designates this function as a subcommand. The name of the subcommand will be taken from the name of the function.

Reviewer: ```suggestion
First we import the `cli` object from `milc`. This is how we interact with the user and control the script's behavior. We use `@cli.argument()` to define a command line flag, `--name`. This also creates a configuration variable named `hello.name` (and the corresponding `user.name`), which the user can set so they don't have to specify the argument. The `cli.subcommand()` decorator designates this function as a subcommand. The name of the subcommand will be taken from the name of the function.
```

# Overview

The QMK CLI operates using the subcommand pattern made famous by git. The main `qmk` script is simply there to setup the environment and pick the correct entrypoint to run.

There are actually 2 different `qmk` programs, but they operate in very similar ways. There is an installable `qmk`, also referred to as the global CLI, which can operate on many different `qmk_firmware` repositories. There is also the bundled `bin/qmk`, also referred to the local CLI, which always operates on the `qmk_firmware` it is located in. In both cases the `qmk` script prepares the environment, loads all of the subcommand modules, and then executes the entrypoint designated by the user's CLI command.

Each subcommand is a self-contained module with an entrypoint (decorated by `@cli.subcommand()`) that performs some action and returns a shell returncode, or None.

# Subcommands

[MILC](https://github.com/clueboard/milc) is the CLI framework `qmk` uses to handle argument parsing, configuration, logging, and many other features. It lets you focus on writing your tool without wasting your time writing glue code.

> Subcomands in the local CLI are always found in `qmk_firmware/lib/python/qmk/cli`. Subcommands in the global CLI are always found in `qmk_cli/qmk_cli/subcommands`. Most subcommands should be located in the local CLI. The global CLI subcommands should be limited to those few needed to setup a `qmk_firmware` environment.

Reviewer 2: ```suggestion
Subcommands in the local CLI are always found in `qmk_firmware/lib/python/qmk/cli`. Subcommands in the global CLI are always found in `qmk_cli/qmk_cli/subcommands`. Most subcommands should be located in the local CLI. The global CLI subcommands should be limited to those few needed to setup a `qmk_firmware` environment.
```

> Subcommands in the local CLI are always found in `qmk_firmware/lib/python/qmk/cli`. Subcommands in the global CLI are always found in `qmk_cli/qmk_cli/subcommands`. Most subcommands should be located in the local CLI. The global CLI subcommands should be limited to those few needed to setup a `qmk_firmware` environment.

Reviewer 3: the global vs local thing is a bit confusing. Which one should developer prioritize? what kinds of things should be local? Should it be moved to an appendix?

Let's start by looking at an example subcommand. This is `lib/python/qmk/cli/hello.py`:

```python
"""QMK Python Hello World

This is an example QMK CLI script.
"""
from milc import cli

@cli.argument('-n', '--name', default='World', help='Name to greet.')
@cli.subcommand('QMK Hello World.')
def hello(cli):
    """Log a friendly greeting.
    """
    cli.log.info('Hello, %s!', cli.config.hello.name)
```

> The first thing we do is import the `cli` object from `milc`. This is how we interact with the user and control the script's behavior. We use `@cli.arguent()` to define a command line flag, `--name`. This also creates a configuration variable named `hello.name` (and the corresponding `default.name`) which the user can set so they don't have to specify the argument. The `cli.subcomand()` decorator designates this function as a subcommand. The name of the subcommand will be taken from the name of the function.

Reviewer 2: s/arguent()/argument()
s/subcomand()/subcommand()

```suggestion
The first thing we do is import the `cli` object from `milc`. This is how we interact with the user and control the script's behavior. We use `@cli.argument()` to define a command line flag, `--name`. This also creates a configuration variable named `hello.name` (and the corresponding `default.name`) which the user can set so they don't have to specify the argument. The `cli.subcommand()` decorator designates this function as a subcommand. The name of the subcommand will be taken from the name of the function.
```

| Blue | {bg_blue} | {bg_lightblue_ex} | {fg_blue} | {fg_lightblue_ex} |
| Cyan | {bg_cyan} | {bg_lightcyan_ex} | {fg_cyan} | {fg_lightcyan_ex} |
| Green | {bg_green} | {bg_lightgreen_ex} | {fg_green} | {fg_lightgreen_ex} |
| Magenta | {bg_magenta} | {bg_lightmagenta_ex} | {fg_magenta} | {fg_lightmagenta_ex} |
| Red | {bg_red} | {bg_lightred_ex} | {fg_red} | {fg_lightred_ex} |
| White | {bg_white} | {bg_lightwhite_ex} | {fg_white} | {fg_lightwhite_ex} |
| Yellow | {bg_yellow} | {bg_lightyellow_ex} | {fg_yellow} | {fg_lightyellow_ex} |

There are also control sequences that can be used to change the behavior of
ANSI output:

| Control Sequences | Description |
|-------------------|-------------|
| {style_bright} | Make the text brighter |
| {style_dim} | Make the text dimmer |
| {style_normal} | Make the text normal (neither `{style_bright}` nor `{style_dim}`) |
| {style_reset_all} | Reset all text attributes to default. (This is automatically added to the end of every string.) |
| {bg_reset} | Reset the background color to the user's default |
| {fg_reset} | Reset the foreground color to the user's default |

> ### Logging (`cli.log`)

Reviewer 3: I would move this table above the formatting

| cli.log.debug | `{fg_cyan}☐{style_reset_all}`
| cli.log.notset | `{style_reset_all}¯\\_(o_o)_/¯`

### Printing (`cli.echo`)

Sometimes you simply need to print text outside of the log system. This is appropriate if you are outputting fixed data or writing out something that should never be logged. Most of the time you should prefer `cli.log.info()` over `cli.echo`.

# Arguments and Configuration

QMK handles the details of argument parsing and configuration for you. When you add a new argument it is automatically incorporated into the config tree based on your subcommand's name and the long name of the argument. You can access this configuration in `cli.config`, using either attribute-style access (`cli.config.<subcommand>.<argument>`) or dictionary-style access (`cli.config['<subcommand>']['<argument>']`).

Under the hood QMK uses [ConfigParser](https://docs.python.org/3/library/configparser.html) to store configurations. This gives us an easy and straightforward way to represent the configuration in a human-editable way. We have wrapped access to this configuration to provide some nicities that ConfigParser does not normally have.

## Reading Configuration Values

> You can interact with `cli.config` in all the ways you'd normally expect. For example the `qmk compile` comand gets the keyboard name from `cli.config.compile.keyboard`. It does not need to know whether that value came from the command line, an environment variable, or the configuration file.

Reviewer 2: ```suggestion
You can interact with `cli.config` in all the ways you'd normally expect. For example the `qmk compile` command gets the keyboard name from `cli.config.compile.keyboard`. It does not need to know whether that value came from the command line, an environment variable, or the configuration file.
```
